### ISSUE TITLE
[Che] Sync broken on Che, cwctl expecting access to a keychain

### ISSUE THREAD

**Issue by johnmcollier** [eclipse/codewind#1446]
**Codewind version:** 0.7.0

**Che version:** 7.3.1
**IDE extension version:** 0.7.0
**Kubernetes cluster:** OCP4

**Description:**
Project updates / sync appears to be broken on Che. The project will deploy fine the first time around, but subsequent updates are not detected. The following error is shown in the logs:
```
[2019-12-09 21:49:26.163] [586.845] Successfully ran installer command: [ project] [ sync] [ -p] [ /projects/gotest] [ -i] [ 7f350360-1acd-11ea-8b3e-29e55b65828b] [ -t] [ 1575928050077]
[2019-12-09 21:49:26.163] [586.845] Output:error dispatching request  {"error":"tx_nopassword","error_description":"Unable to find password in keychain"}
Status: Unable to find password in keychain

[2019-12-09 21:49:26.163] [586.845] Updating timestamp to latest: 1575928166078
127.0.0.1 - - [09/Dec/2019:21:49:28 +0000] "GET /socket.io/?EIO=3&transport=polling&t=Mxienud&b64=1&sid=oGPP-2YEWVYpm4PNAAAA HTTP/1.1" 200 3 "-" "node-XMLHttpRequest" "-"
127.0.0.1 - - [09/Dec/2019:21:49:28 +0000] "POST /socket.io/?EIO=3&transport=polling&t=Mxiet_G&b64=1&sid=oGPP-2YEWVYpm4PNAAAA HTTP/1.1" 200 2 "-" "node-XMLHttpRequest" "-"
```

It looks like `cwctl` is trying to retrieve a non-existent password from a keychain, but since cwctl is running inside the Che sidecar, there is no keychain setup, nor is there no password to store in the keychain (cwctl is calling PFE directly, not through Gatekeeper like on Hybrid)

**Steps to reproduce:**
1. Create a Che 0.7.0 workspace, use this [devfile](https://raw.githubusercontent.com/eclipse/codewind-che-plugin/0.7.0/devfiles/0.7.0/devfile.yaml)
2. Deploy some project. E.g. Go or Microprofile
3. Make a code change. 
4. The project will **not** rebuild, the codewind-che-sidecar logs will show the error I posted above

**Comment by johnmcollier** [comment]
It looks like the issue is here: https://github.com/eclipse/codewind-installer/blob/0.7.0/pkg/project/sync.go#L217

```
resp, httpSecError := sechttp.DispatchHTTPRequest(client, req, conInfo)
```

When doing a file upload during sync, it's calling out to `sechttp.DispatchHTTPRequest`, which then calls out to the go-keyring service. Since this is running on Che, and no credentials to store in a keyring, I'm wondering if the fix is to just modify `DispatchHTTPRequest` to work without the keyring when running on Che

**Comment by markcor11** [comment]
sechttp only uses the keychain when the connection ID is not equal to “local”, For “local” projects a default unauthenticated request is made.

**Comment by johnmcollier** [comment]
@markcor11 I took at look in the connections .json files
- I had one project in my workspace, but three connections .json file, is that correct? Only one matched the project ID of the project
- All of the connections .json files had `"connectionID": "local"`, so it is getting set properly.

**Comment by markcor11** [comment]
@johnmcollier The project ID is generated by PFE and is PFE instance specific.  A single project might have multiple files in `.codewind/config/connections` if it has been deployed on additional "remotes" but only one of them will be named the `{projectID}.json`. You found the matching one,  and that does look correct being set in CHE to "local".  This then suggests that either somewhere in the sync command the ID has got lost, or the initial request to PFE "local" failed and returned an error.  

There is a lot of trace debug for sechttp which can be enabled with the global cwctl flag `--loglevel trace`.  It would be interesting to see the output within that pod of  : 

```
cwctl --insecure --loglevel trace project sync -p /projects/gotest -i 7f350360-1acd-11ea-8b3e-29e55b65828b -t 1575928050077
```

This will likely show an error for the initial http request to local failed.  If it does fail,  cwctl will revert to trying to connect to PFE three different ways: using access tokens from the keychain, a refresh token from the keychain, or credentials from the keychain. 

Next steps :

- [ ] Me:  Attempt to reproduce with latest
- [ ] Me:  Debug

**Comment by markcor11** [comment]
Unable to reproduce this.

Created a new go app, modified it, watched it rebuild, status changed to running, refreshed webpage and app change appeared. 

```

10/12/19 12:35:19 Turbine] [INFO] 14e388c0-1b49-11ea-9c40-d54b3f362761: Project 14e388c0-1b49-11ea-9c40-d54b3f362761 file changed
--
  | [10/12/19 12:35:19 mark-go] [INFO] Detected other file changes, Codewind will build the project
  | [10/12/19 12:35:19 mark-go] [INFO] File change detected. Project will re-build.
  | [10/12/19 12:35:19 mark-go] [INFO] projectInfo.deploymentRegistry: undefined
  | [10/12/19 12:35:19 mark-go] [INFO] Image Push Registry: docker.io/marko11
  | [10/12/19 12:35:19 mark-go] [INFO] The projectInfo has been updated for imagePushRegistry: {"projectID":"14e388c0-1b.....
```

```

[10/12/19 12:35:19 mark-go] [INFO] Emitting event
--
  | message: projectStatusChanged
  | data: {
  | "projectID": "14e388c0-1b49-11ea-9c40-d54b3f362761",
  | "buildStatus": "inProgress",
  | "detailedBuildStatus": "Creating image"
  | }
```

**Comment by johnmcollier** [comment]
@markcor11 Hmm, let me try again. I was able to reproduce on 0.7.0 and master yesterday. Unless it took a while for the image to get refreshed with new cwctl binaries? I'll test again after the playback call today.

**Comment by johnmcollier** [comment]
@markcor11 I can still reproduce on Che with the 0.7.0 devfile, how did you test with Che?

**Comment by johnmcollier** [comment]
```
/projects/.codewind/config/connections $ cwctl --insecure --loglevel trace proje
ct sync -p /projects/gotest -i 0406b850-1b59-11ea-ad56-77ccf4d8db24 -t 0
TRAC[0000] Request URL: PUT https://localhost:9090/api/v1/projects/0406b850-1b59-11ea-ad56-77ccf4d8db24/upload
TRAC[0000] Received HTTP Status code: 200
TRAC[0000] Request URL: PUT https://localhost:9090/api/v1/projects/0406b850-1b59-11ea-ad56-77ccf4d8db24/upload
TRAC[0000] Received HTTP Status code: 200
TRAC[0000] Request URL: PUT https://localhost:9090/api/v1/projects/0406b850-1b59-11ea-ad56-77ccf4d8db24/upload
TRAC[0000] Received HTTP Status code: 200
TRAC[0000] Request URL: PUT https://localhost:9090/api/v1/projects/0406b850-1b59-11ea-ad56-77ccf4d8db24/upload
TRAC[0000] Received HTTP Status code: 200
TRAC[0000] Request URL: PUT https://localhost:9090/api/v1/projects/0406b850-1b59-11ea-ad56-77ccf4d8db24/upload
TRAC[0000] Received HTTP Status code: 200
TRAC[0000] Request URL: PUT https://localhost:9090/api/v1/projects/0406b850-1b59-11ea-ad56-77ccf4d8db24/upload
TRAC[0000] Received HTTP Status code: 200
TRAC[0000] Request URL: PUT https://localhost:9090/api/v1/projects/0406b850-1b59-11ea-ad56-77ccf4d8db24/upload
TRAC[0000] Received HTTP Status code: 200
TRAC[0000] Request URL: PUT https://localhost:9090/api/v1/projects/0406b850-1b59-11ea-ad56-77ccf4d8db24/upload
TRAC[0000] Received HTTP Status code: 200
TRAC[0000] Request URL: POST /api/v1/projects/0406b850-1b59-11ea-ad56-77ccf4d8db24/upload/end
TRAC[0000] sendRequest: REQUEST FAILED
TRAC[0000] Getting Connection: local
TRAC[0000] Retrieving an access token from the keychain
TRAC[0000] Access token not found in keychain
TRAC[0000] Retrieving a refresh token from the keychain
TRAC[0000] Refresh token not found in keychain
TRAC[0000] Re-authenticate using cached credentials from the keychain
TRAC[0000] ERROR:  exec: "dbus-launch": executable file not found in $PATH
error dispatching request  {"error":"tx_nopassword","error_description":"Unable to find password in keychain"}
Status: Unable to find password in keychain
```

**Comment by markcor11** [comment]
just reproduced it : 

```
TRAC[0001] Received HTTP Status code: 200               
TRAC[0001] Request URL: POST /api/v1/projects/14e388c0-1b49-11ea-9c40-d54b3f362761/upload/end 
TRAC[0001] sendRequest: REQUEST FAILED                  
TRAC[0001] Getting Connection: local                    
TRAC[0001] Retrieving an access token from the keychain 
TRAC[0001] Access token not found in keychain           
TRAC[0001] Retrieving a refresh token from the keychain 
TRAC[0001] Refresh token not found in keychain          
TRAC[0001] Re-authenticate using cached credentials from the keychain 
TRAC[0001] ERROR:  exec: "dbus-launch": executable file not found in $PATH 
error dispatching request  {"error":"tx_nopassword","error_description":"Unable to find password in keychain"}
Status: Unable to find password in keychain
```

**Comment by markcor11** [comment]
```
TRAC[0001] Request URL: POST /api/v1/projects/14e388c0-1b49-11ea-9c40-d54b3f362761/upload/end 
TRAC[0001] sendRequest: REQUEST FAILED     
```

Seems the hostname is missing....

**Comment by markcor11** [comment]
The bug is here : https://github.com/eclipse/codewind-installer/blob/a1d31b08fbf4509a819f2b5d20892f2dbd9887a4/pkg/project/sync.go#L205

because the connection info for local is never pre-populated (local desktop mode PFE starts up on a random port so we dont persist it in connections.json). The code change should be to use the URL obtained from the PFEOriginFromConnection function (which is both docker and CHE aware)

eg : 

```
conURL, conURLErr := config.PFEOriginFromConnection(conInfo)
```

**Comment by malincoln** [comment]
@johnmcollier pls verify and close.  Thanks

**Comment by johnmcollier** [comment]
Verifying, fixed.

Thanks @markcor11!
